### PR TITLE
Add Geb Backend Evaluator with some extra subcommands

### PR DESCRIPTION
**Pull request by jonaprieto**
This PR introduces an evaluator for the Geb STLC interface/fragment and other related commands, including a REPL to interact with his backend.

- https://github.com/anoma/geb/blob/mariari/binaries/src/specs/lambda.lisp

We have included a  REPL and support for commands such as read and eval here. Check out:

```
juvix dev geb --help
```

- [x] Add Geb evaluator with the two basic eval strategies.
- [x] Add quasi quoter: return morphisms from typed geb values.
- [x] Add type/object inference for morphisms.
- [x] All combined: morphisms-eval-to-morphisms
- [x] Parse and pretty printer Geb values (without quoting them)
- [x] Parse files containing Geb terms:
  - [x] Saved in a .lisp file according to anoma/geb example (typed object).
  - [x] Store in a .geb file simple as simple lisp expression.
- [x] Add related commands to the CLI for `dev geb`:
  - [x] Subcommand: eval
  - [x] Subcommand: read
  - [x] Subcommand: infer
  - [x] Subcommand: repl
  - [x] Subcommand: check 
- [x] Minor changes `hom` by `!->` in the Geb prettyprinter
- [x] Add tests for:
   - [x] New subcommand (smoke tests)
   - [x] Eval

Issues to solve after merging this PR: 

- Add location to Geb ast for proper error location.
- Add tests for all related subcommands, e.g. check, and infer.
- Check compilation from Core to Geb: (run inferObject with the type provided by the core node).
- [x] Update the vs code-plugin to load Geb repl and eval. (https://github.com/anoma/vscode-juvix/commit/31994c8684e9b44a62d58ac13304b9e04503410c)

### CODE REVIEW

File: foo.lisp [hunk 1]
```diff
@@ -0,0 +1,43 @@
+(defpackage #:foo
```
Comment (lukaszcz) [inline foo.lisp:1]: Remove this file

File: tests/Geb/positive/Core/test003.jvc [hunk 2]
```diff
@@ -11,4 +11,4 @@ type enum {
   opt1 b := b;
   opt2 b f := f b;
   opt3 b1 f b2 := f b1 b2;
-}) (opt3 true (\(x : bool) \(y : bool) if y then false else x) false)
+}) (opt3 true (\(x : bool) \(y : bool) if y then false else x) false)
```
Comment (lukaszcz) [inline tests/Geb/positive/Core/test003.jvc:14]: Why remove the newline?

File: tests/Geb/positive/Core/test004.jvc [hunk 3]
```diff
@@ -2,4 +2,4 @@
 def not : bool -> bool := \(x : bool) if x then false else true;
 def and : bool -> bool -> bool := \(x : bool) \(y : bool) if x then y else false;
 
-and (not false) (not (not true))
+and (not false) (not (not true))
```
Comment (lukaszcz) [inline tests/Geb/positive/Core/test004.jvc:5]: Why remove the newline?

File: tests/Geb/positive/test001.lisp [hunk 4]
```diff
@@ -0,0 +1,49 @@
+(defpackage #:test001
```
Comment (lukaszcz) [inline tests/Geb/positive/test001.lisp:1]: I don't think we should keep two or more versions of every test. There should be either test001.geb or test001.lisp, but not both. Maintaining them -- keeping in sync -- will become an unnecessary burden, and the only purpose of it is to check that we're parsing the lisp header correctly, which is not even strictly necessary.

File: app/Commands/Dev/Geb/Repl.hs [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+type ReplS = State.StateT ReplState IO
+
+type Repl a = Repline.HaskelineT ReplS a
+
+newtype ReplContext = ReplContext
+  { _replContextEntryPoint :: EntryPoint
+  }
+
+data ReplState = ReplState
+  { _replStateContext :: Maybe ReplContext,
```
Comment (paulcadman) [inline app/Commands/Dev/Geb/Repl.hs:29]: This can be `Maybe EntryPoint` to remove one layer of indirection.

File: foo.lisp (hunk 1 above)
Comment (jonaprieto) [reply foo.lisp]: Done.

File: app/Commands/Dev/Geb/Eval/Options.hs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+      "call-by-name" -> Right Geb.CallByName
+      s -> Left $ "Unrecognised evaluation strategy: " <> s
+
+    customShow :: Geb.EvalStrategy -> String
+    customShow = \case
+      Geb.CallByValue -> "call-by-value"
+      Geb.CallByName -> "call-by-name"
+
+optNormalise :: Parser Bool
+optNormalise =
```
Comment (lukaszcz) [inline app/Commands/Dev/Geb/Eval/Options.hs:57]: We're not doing normalisation, only evaluation. What is this option supposed to turn on?

File: app/Commands/Dev/Geb/Options.hs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+import Commands.Dev.Geb.Read.Options
+import Commands.Dev.Geb.Repl.Options
+import CommonOptions
+
+data GebCommand
+  = GebCommandRepl GebReplOptions
+  | GebCommandEval GebEvalOptions
+  | GebCommandRead GebReadOptions
+  | GebCommandInfer GebInferOptions
+  | GebCommandCheck GebInferOptions -- Same as infer for now
```
Comment (lukaszcz) [inline app/Commands/Dev/Geb/Options.hs:19]: Remove comment (doesn't contribute new essential info)

File: app/Commands/Extra/Compile/Options.hs [hunk 8]
```diff
@@ -57,8 +57,7 @@ parseCompileOptions = do
       )
   _compileTerm <-
     switch
-      ( short 'T'
-          <> long "only-term"
+      ( long "only-term"
```
Comment (lukaszcz) [inline app/Commands/Extra/Compile/Options.hs:60]: Why remove the short '-T' version? I think it's handy and it fits with the other short upper-case options which modify the exact output format for other backend (-E, -S).

File: examples/demo/Demo.juvix [hunk 9]
```diff
@@ -32,29 +32,28 @@ module Demo;
```
Comment (lukaszcz) [inline examples/demo/Demo.juvix:32]: Revert the formatting changes to *.juvix files (they go into a separate PR)

File: src/Juvix/Compiler/Backend/Geb/Evaluator.hs [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+      mapError (JuvixError @CheckingError) (inferObject m)
+  if
+      | requiresObjectInfo val -> quote (Just obj) val
+      | otherwise -> quote Nothing val
+
+type EvalEffects r = Members '[Reader Env, Error EvalError] r
+
+eval :: EvalEffects r => Morphism -> Sem r GebValue
+eval morph = do
+  -- env <- asks (^. envContext)
```
Comment (lukaszcz) [inline src/Juvix/Compiler/Backend/Geb/Evaluator.hs:71]: Remove unnecessary comment

File: src/Juvix/Compiler/Backend/Geb/Evaluator.hs [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+
+evalApp :: EvalEffects r => Application -> Sem r GebValue
+evalApp app = do
+  evalStrategy <- asks (^. envEvaluatorOptions . evaluatorOptionsEvalStrategy)
+  let maybeForce :: GebValue -> GebValue
+      maybeForce = case evalStrategy of
+        CallByName -> id
+        CallByValue -> force
+  arg <- maybeForce <$> eval (app ^. applicationRight)
+  apply (app ^. applicationLeft) arg
```
Comment (lukaszcz) [inline src/Juvix/Compiler/Backend/Geb/Evaluator.hs:158]: I don't think this gives you the evaluation strategy you want, because for `force` to be evaluated, `arg` needs to be evaluated, and it won't be until you force its evaluation, e.g., by pattern matching on it.

Besides, I don't think you need `deepseq` to achieve call-by-value. Call-by-value can be achieved just by forcing shallow evaluation of all function, constructor and primitive arguments. Because if you consistently evaluate all arguments, then anything contained in some GebValue will already be evaluated because it got forced earlier..

Comment (lukaszcz) [reply src/Juvix/Compiler/Backend/Geb/Evaluator.hs:158]: And we don't really need those evaluation strategies. In STLC the result is always the same anyway. I doubt it would ever have any visible impact on performance in our usage.

File: src/Juvix/Compiler/Backend/Geb/Evaluator.hs [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+
+apply ::
+  EvalEffects r =>
+  Morphism ->
+  GebValue ->
+  Sem r GebValue
+apply fun' arg = do
+  -- env <- asks (^. envContext)
```
Comment (lukaszcz) [inline src/Juvix/Compiler/Backend/Geb/Evaluator.hs:166]: Remove comment

File: src/Juvix/Compiler/Backend/Geb/Evaluator.hs [hunk 13]
```diff
@@ -0,0 +1,377 @@
...
+  -- trace
+  --   ( "apply\n"
+  --       <> (" fun:= " <> show fun' <> "\n")
+  --       <> (" arg:= " <> show arg <> "\n")
+  --       <> (" env:=" <> show env <> "\n")
+  --   )  $ do
+  fun <- eval fun'
+  case fun of
+    GebValueClosure cls ->
+      -- trace ("cls:= " <> show cls <> "\n") $
```
Comment (lukaszcz) [inline src/Juvix/Compiler/Backend/Geb/Evaluator.hs:176]: Remove comment

File: src/Juvix/Compiler/Backend/Geb/Evaluator.hs [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+    EvalError
+      { _evalErrorMsg = "Absurd can not be evaluated.",
+        _evalErrorGebValue = Nothing,
+        _evalErrorGebExpression = Just morph
+      }
+
+evalPair :: EvalEffects r => Pair -> Sem r GebValue
+evalPair pair = do
+  left <- eval $ pair ^. pairLeft
+  right <- eval $ pair ^. pairRight
```
Comment (lukaszcz) [inline src/Juvix/Compiler/Backend/Geb/Evaluator.hs:116]: If you want call-by-value, you also need to force evaluation of `left` and `right` here. Analogously for all other constructors and primitives.

File: src/Juvix/Compiler/Backend/Geb/Evaluator.hs [hunk 15]
```diff
@@ -0,0 +1,377 @@
...
+            _evalErrorGebExpression = Just (MorphismCase c)
+          }
+
+evalBinop ::
+  Members '[Reader Env, Error EvalError] r =>
+  Binop ->
+  Sem r GebValue
+evalBinop binop = do
+  left <- eval $ binop ^. binopLeft
+  right <- eval $ binop ^. binopRight
```
Comment (lukaszcz) [inline src/Juvix/Compiler/Backend/Geb/Evaluator.hs:220]: For call-by-value, `left` and `right` need to be forced here.

File: src/Juvix/Compiler/Backend/Geb/Evaluator.hs [hunk 16]
```diff
@@ -0,0 +1,377 @@
...
+quoteClosure ty cls = do
+  case ty of
+    Just (ObjectHom funObj) ->
+      return $
+        MorphismLambda
+          Lambda
+            { _lambdaVarType = funObj ^. homDomain,
+              _lambdaBodyType = funObj ^. homCodomain,
+              _lambdaBody =
+                cls ^. valueClosureLambdaBody
```
Comment (lukaszcz) [inline src/Juvix/Compiler/Backend/Geb/Evaluator.hs:311]: This is incorrect. You need to substitute the free variables of the body for the values in the environment.

File: src/Juvix/Compiler/Backend/Geb/Evaluator/Data/Values.hs [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+instance HasAtomicity GebValue where
+  atomicity = \case
+    GebValueMorphismInteger {} -> Atom
+    GebValueMorphismLeft {} -> Aggregate appFixity
+    GebValueMorphismPair {} -> Aggregate appFixity
+    GebValueMorphismRight {} -> Aggregate appFixity
+    GebValueMorphismUnit -> Atom
+    GebValueClosure {} -> Aggregate appFixity
+
+instance Show GebValue where
```
Comment (lukaszcz) [inline src/Juvix/Compiler/Backend/Geb/Evaluator/Data/Values.hs:62]: I think all those `Show` instances should be removed. They duplicate the pretty printing functionality in a slightly different syntax. I don't think the fact that don't print the types is enough justification to keep them.

File: src/Juvix/Compiler/Backend/Geb/Language.hs [hunk 18]
```diff
@@ -144,6 +216,74 @@ instance HasAtomicity Object where
...
 
+instance HasAtomicity Expression where
+  atomicity = \case
+    ExpressionMorphism m -> atomicity m
+    ExpressionObject o -> atomicity o
+
+instance HasAtomicity TypedMorphism where
+  atomicity _ = Aggregate appFixity
+
+instance Show Morphism where
```
Comment (lukaszcz) [inline src/Juvix/Compiler/Backend/Geb/Language.hs:227]: I think the `Show` instances should be removed, because they duplicate the functionality of pretty printing.

File: package.yaml [hunk 19]
```diff
@@ -33,6 +33,7 @@ dependencies:
 - blaze-html == 0.9.*
 - bytestring == 0.11.*
 - containers == 0.6.*
+- deepseq == 1.4.*
```
Comment (lukaszcz) [inline package.yaml:36]: I don't think we really need this dependency. See comments in the evaluator file.

File: app/Commands/Dev/Geb/Options.hs [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+    inferInfo =
+      info
+        (GebCommandInfer <$> parseGebInferOptions)
+        (progDesc "Infer the GebObject for a Geb morphism found in the given file. ")
+
+    checkInfo :: ParserInfo GebCommand
+    checkInfo =
+      info
+        (GebCommandInfer <$> parseGebInferOptions)
+        (progDesc "Check the GebObject provided matches th given Geb morphism. ")
```
Comment (lukaszcz) [inline app/Commands/Dev/Geb/Options.hs:76]: "Check the given Geb object matches the given Geb morphism."

File: app/Commands/Dev/Geb/Options.hs [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+    readInfo =
+      info
+        (GebCommandRead <$> parseGebReadOptions)
+        (progDesc "Read a JuvixGeb file and pretty print it")
+
+    inferInfo :: ParserInfo GebCommand
+    inferInfo =
+      info
+        (GebCommandInfer <$> parseGebInferOptions)
+        (progDesc "Infer the GebObject for a Geb morphism found in the given file. ")
```
Comment (lukaszcz) [inline app/Commands/Dev/Geb/Options.hs:70]: GebObject -> Geb object

Comment (lukaszcz) [inline app/Commands/Dev/Geb/Options.hs:70]: GebObject -> Geb object

File: src/Juvix/Compiler/Backend/Geb/Pretty/Values/Base.hs [hunk 22]
```diff
@@ -0,0 +1,97 @@
+module Juvix.Compiler.Backend.Geb.Pretty.Values.Base where
```
Comment (lukaszcz) [inline src/Juvix/Compiler/Backend/Geb/Pretty/Values/Base.hs:1]: Wouldn't it be better if we just added the instances from this file to Pretty/Base.hs? This would allow us to delete Pretty/Values.hs and avoid much code duplication.

File: app/Commands/Dev/Geb/Repl.hs [hunk 23]
```diff
@@ -0,0 +1,323 @@
...
+    Right _ -> printError (error "Inference only works on Geb morphisms.")
+
+checkTypedMorphism :: String -> Repl ()
+checkTypedMorphism gebMorphism = Repline.dontCrash $ do
+  case Geb.runParser' replPath (pack gebMorphism) of
+    Left err -> printError (JuvixError err)
+    Right tyMorphism@(Geb.TypedMorphism {}) -> do
+      case run . runError @CheckingError $ Geb.check' tyMorphism of
+        Right obj -> renderOut (Geb.ppOut Geb.defaultEvaluatorOptions obj)
+        Left err -> undefined -- printError (JuvixError err)
```
[the remark this answers is not in the document]
Comment (jonaprieto) [reply app/Commands/Dev/Geb/Repl.hs]: Done

File: src/Juvix/Compiler/Core/Data/BinderList.hs [hunk 24]
```diff
@@ -8,12 +10,18 @@ data BinderList a = BinderList
...
+
+instance NFData a => NFData (BinderList a)
 
 makeLenses ''BinderList
 
 fromList :: [a] -> BinderList a
 fromList l = BinderList (length l) l
 
+empty :: BinderList a
+empty = fromList []
```
Comment (lukaszcz) [inline src/Juvix/Compiler/Core/Data/BinderList.hs:23]: There is `mempty` already. This should not be duplicated with `empty`.

File: test/BackendGeb/Eval/Negative.hs [hunk 25]
```diff
@@ -0,0 +1,32 @@
+module BackendGeb.Eval.Negative where
```
Comment (lukaszcz) [inline test/BackendGeb/Eval/Negative.hs:1]: Since we don't have any negative tests, I think this file should be removed. It contains only dead code.

File: tests/Geb/positive/Core/test006.jvc [hunk 26]
```diff
@@ -18,4 +18,4 @@ func (y / x) + -- 17 div 5 + 4 = 7
 (z * x + y) + -- 17
 (vx + vy * (z + 1)) + -- 37
 f (h g 2 3) 4 -- -29
--- result: 32
+-- result: 32
```
[the remark this answers is not in the document]
Comment (jonaprieto) [reply tests/Geb/positive/Core/test006.jvc]: Done

File: tests/Geb/positive/test001.lisp (hunk 4 above)
Comment (jonaprieto) [reply tests/Geb/positive/test001.lisp]: Done.

File: app/Commands/Dev/Geb/Repl.hs (hunk 5 above)
Comment (jonaprieto) [reply app/Commands/Dev/Geb/Repl.hs]: Done.

File: tests/Geb/positive/Core/test005.jvc [hunk 27]
```diff
@@ -1,3 +1,3 @@
 -- basic arithmetic
 
-5 + 2 * 3
+5 + 2 * 3
```
Comment (lukaszcz) [inline tests/Geb/positive/Core/test005.jvc:3]: Avoid changes that don't really change anything. Revert to the version with the newline.

File: tests/Geb/positive/Core/test002.jvc [hunk 28]
```diff
@@ -4,4 +4,9 @@ type optbool {
   Nothing : optbool;
 };
 
-(\(x : bool) \(o : optbool) { case o of { Just b := if x then true else b; Nothing := false } }) false (Just true)
+(\(x : bool) \(o : optbool) { 
+case o of { 
+  Just b := if x then true else b; 
+  Nothing := false } }) 
+  false 
+  (Just true)
```
Comment (lukaszcz) [inline tests/Geb/positive/Core/test002.jvc:12]: This formatting change actually breaks the formatting, because the `case` has the wrong indentation.

File: tests/positive/MiniHaskell/HelloWorld.juvix [hunk 29]
```diff
@@ -8,7 +8,6 @@ type V :=
   zeroV : V |
```
Comment (lukaszcz) [inline tests/positive/MiniHaskell/HelloWorld.juvix:8]: Revert formatting changes in *.juvix files.

File: src/Juvix/Compiler/Backend/Geb/Analysis/TypeChecking.hs [hunk 30]
```diff
@@ -0,0 +1,283 @@
...
+type InferEffects r = Members '[Reader InferenceEnv, Error CheckingError] r
+
+inferObject ::
+  Members '[Reader InferenceEnv, Error CheckingError] r =>
+  Morphism ->
+  Sem r Object
+inferObject = \case
+  MorphismUnit -> return ObjectTerminal
+  MorphismInteger {} -> return ObjectInteger
+  MorphismAbsurd x -> inferObject x
```
Comment (lukaszcz) [inline src/Juvix/Compiler/Backend/Geb/Analysis/TypeChecking.hs:58]: This is incorrect. The type depends on the context, and cannot be inferred just based on the Absurd term if it's not provided in the term (see my comment in https://github.com/anoma/geb/issues/53)

File: src/Juvix/Compiler/Backend/Geb/Analysis/TypeChecking.hs [hunk 31]
```diff
@@ -0,0 +1,283 @@
...
+      checkListSameType args ObjectInteger
+      return outTy
+    OpDiv -> do
+      checkListSameType args ObjectInteger
+      return outTy
+    OpMod -> do
+      checkListSameType args ObjectInteger
+      return outTy
+    OpEq -> do
+      checkSameType args
```
Comment (lukaszcz) [inline src/Juvix/Compiler/Backend/Geb/Analysis/TypeChecking.hs:198]: It is yet unspecified what the types of arguments to Eq can be (depends on GEB - Eq is not yet there), but probably only integers, maybe some other primitive types.

File: src/Juvix/Compiler/Backend/Geb/Evaluator/Data/Values.hs (hunk 17 above)
Comment (jonaprieto) [reply src/Juvix/Compiler/Backend/Geb/Evaluator/Data/Values.hs]: This was for debugging purposes. Removed.

File: src/Juvix/Compiler/Backend/Geb/Language.hs (hunk 18 above)
Comment (jonaprieto) [reply src/Juvix/Compiler/Backend/Geb/Language.hs]: This was for debugging purposes. Removed.

File: package.yaml (hunk 19 above)
Comment (jonaprieto) [reply package.yaml]: Done.

File: app/Commands/Dev/Geb/Options.hs (hunk 20 above)
Comment (jonaprieto) [reply app/Commands/Dev/Geb/Options.hs]: Done.

File: app/Commands/Dev/Geb/Options.hs (hunk 21 above)
Comment (jonaprieto) [reply app/Commands/Dev/Geb/Options.hs]: Done.

Comment (jonaprieto) [reply app/Commands/Dev/Geb/Options.hs]: Done.

File: src/Juvix/Compiler/Backend/Geb/Pretty/Values/Base.hs (hunk 22 above)
Comment (jonaprieto) [reply src/Juvix/Compiler/Backend/Geb/Pretty/Values/Base.hs]: Done.

File: src/Juvix/Compiler/Core/Data/BinderList.hs (hunk 24 above)
Comment (jonaprieto) [reply src/Juvix/Compiler/Core/Data/BinderList.hs]: Done.

File: test/BackendGeb/Eval/Negative.hs (hunk 25 above)
Comment (jonaprieto) [reply test/BackendGeb/Eval/Negative.hs]: Done.

[the remark this answers is not in the document]
Comment (jonaprieto) [reply test/BackendGeb/Eval/Negative.hs]: Done.

File: app/Commands/Dev/Geb/Eval.hs [hunk 32]
```diff
@@ -0,0 +1,53 @@
...
+    let env :: Geb.Env =
+          Geb.Env
+            { _envEvaluatorOptions = opts',
+              _envContext = mempty
+            }
+    if
+        | opts' ^. Geb.evaluatorOptionsOutputMorphism ->
+            case Geb.evalAndOutputMorphism' env morphism of
+              Left err -> exitJuvixError err
+              Right m -> renderStdOut (Geb.ppOut opts' m)
```
Comment (lukaszcz) [inline app/Commands/Dev/Geb/Eval.hs:48]: I would argue to add `embed (putStrLn "")` here. It's annoying that `juvix dev geb eval` output doesn't end with a newline when you run it in the command line.

File: app/Commands/Dev/Geb/Eval.hs [hunk 33]
```diff
@@ -0,0 +1,53 @@
...
+        | otherwise ->
+            case Geb.eval' env morphism of
+              Left err -> exitJuvixError err
+              Right m -> renderStdOut (GebValue.ppOut opts' m)
```
Comment (lukaszcz) [inline app/Commands/Dev/Geb/Eval.hs:52]: As above.

File: tests/Geb/positive/app-case-on.geb [hunk 34]
```diff
@@ -0,0 +1,22 @@
+(app
```
Comment (lukaszcz) [inline tests/Geb/positive/app-case-on.geb:1]: The types don't match in this file.

File: src/Juvix/Compiler/Backend/Geb/Language.hs (hunk 18 above)
Comment (jonaprieto) [reply src/Juvix/Compiler/Backend/Geb/Language.hs]: Done.

File: package.yaml (hunk 19 above)
Comment (jonaprieto) [reply package.yaml]: Done.

File: app/Commands/Dev/Geb/Options.hs (hunk 20 above)
Comment (jonaprieto) [reply app/Commands/Dev/Geb/Options.hs]: Done.

File: tests/Geb/positive/app-fst-pair.geb [hunk 35]
```diff
@@ -0,0 +1,15 @@
+(app
+    int 
+    int
```
Comment (lukaszcz) [inline tests/Geb/positive/app-fst-pair.geb:3]: This should be (prod int int)?

File: app/Commands/Dev/Geb/Options.hs (hunk 21 above)
Comment (jonaprieto) [reply app/Commands/Dev/Geb/Options.hs]: Done.

File: tests/Geb/positive/app-fst-pair.geb (hunk 35 above)
Comment (lukaszcz) [inline tests/Geb/positive/app-fst-pair.geb:3]: This should be (prod int int).

File: src/Juvix/Compiler/Backend/Geb/Evaluator.hs (hunk 14 above)
Comment (jonaprieto) [reply src/Juvix/Compiler/Backend/Geb/Evaluator.hs:104]: Done.

File: src/Juvix/Compiler/Backend/Geb/Evaluator.hs (hunk 13 above)
Comment (jonaprieto) [reply src/Juvix/Compiler/Backend/Geb/Evaluator.hs]: Done.

File: tests/Geb/positive/app-fst-pair.geb (hunk 35 above)
Comment (jonaprieto) [reply tests/Geb/positive/app-fst-pair.geb:3]: I wrote the example just for the evaluator. So, the types are not meant to be correct.

File: src/Juvix/Compiler/Backend/Geb/Evaluator.hs [hunk 36]
```diff
@@ -0,0 +1,377 @@
...
+requiresObjectInfo :: GebValue -> Bool
+requiresObjectInfo = \case
+  GebValueMorphismUnit -> False
+  GebValueMorphismInteger {} -> False
+  GebValueClosure {} -> True
+  GebValueMorphismLeft {} -> True
+  GebValueMorphismRight {} -> True
+  GebValueMorphismPair {} -> True
+
+quote :: Maybe Object -> GebValue -> Sem r Morphism
```
[the remark this answers is not in the document]
Comment (jonaprieto) [reply src/Juvix/Compiler/Backend/Geb/Evaluator.hs:267]: Sounds good to me.

File: app/Commands/Dev/Geb/Eval.hs (hunk 32 above)
Comment (jonaprieto) [reply app/Commands/Dev/Geb/Eval.hs:48]: Sounds good to me.

**Review by lukaszcz — CHANGES_REQUESTED** [review]
Revert:
* formatting changes to *.juvix files

Make correct or remove (I would advocate the second):
* evaluation strategies,
* quoting of values to morphisms.

File: tests/Geb/positive/app-case-on.geb (hunk 34 above)
Comment (jonaprieto) [reply tests/Geb/positive/app-case-on.geb:1]: On purpose.

File: src/Juvix/Compiler/Backend/Geb/Analysis/TypeChecking.hs [hunk 37]
```diff
@@ -0,0 +1,283 @@
...
+      return outTy
+    OpLt -> do
+      checkSameType args
```
[the remark this answers is not in the document]
Comment (jonaprieto) [reply src/Juvix/Compiler/Backend/Geb/Analysis/TypeChecking.hs:201]: Yes. Good.

File: src/Juvix/Compiler/Backend/Geb/Evaluator.hs [hunk 38]
```diff
@@ -0,0 +1,336 @@
...
+    MorphismRight m -> GebValueMorphismRight <$> eval m
+    MorphismSecond s -> evalSecond s
+    MorphismUnit -> return GebValueMorphismUnit
+    MorphismVar x -> evalVar x
+
+evalVar :: EvalEffects r => Var -> Sem r GebValue
+evalVar var = do
+  ctx <- asks (^. envContext)
+  let val = Context.lookup (var ^. varIndex) ctx
+  -- trace ("varLookup := " <> show val) $
```
Comment (lukaszcz) [inline src/Juvix/Compiler/Backend/Geb/Evaluator.hs:89]: Remove comment

File: src/Juvix/Compiler/Backend/Geb/Pretty/Values.hs [hunk 39]
```diff
@@ -0,0 +1,28 @@
+module Juvix.Compiler.Backend.Geb.Pretty.Values
```
Comment (lukaszcz) [inline src/Juvix/Compiler/Backend/Geb/Pretty/Values.hs:1]: Remove this file. It almost exactly copies Geb/Pretty.hs

**Review by lukaszcz — CHANGES_REQUESTED** [review]
Address the small issues mentioned in the still unresolved conversations.

**Review by lukaszcz — APPROVED** [review]
